ARCs/arc-0004.md

```diff
@@ -0,0 +1,498 @@
+described later.
+
+Invoking a method involves creating an application call transaction to
+specifically call that method. Methods are different from internal
+"helper functions" that may exist in a contract, but are not
+externally callable. In the future, applications may be able to invoke
+methods from other applications.
+
+
+#### Method Signature
+
+A method signature is a unique identifier for a method. The signature
+is a string that consists of the method's name, an open parenthesis, a
+comma-separated list of the types of its arguments, a closing
+parenthesis, and the method's return type, or `void` if it does not
+return a value. The names of arguments are _not_ included in a
+method's signature.
+
+For example, `add(uint64,uint64)uint128` is the method signature for a
+method named `add` which takes two uint64 parameters and returns a
+uint128.  Signatures are encoded in utf-8, and method names may not
+contain an open parenthesis or whitespace. For the benefit of univeral
+interoperability (especially in HLLs), names should be further
+restricted to by the regular expression `[A-Za-z0-9_]+`. The preferred
+naming scheme for methods is camelCase.
+
+
+#### Method Selector
+
+Method signatures contain all the information needed to identify a
+method, however the length of a signature is unbounded. Rather than
+consume program space with such strings, a method selector is used to
+identify methods in calls. A method selector is the first four bytes of
+the SHA-512/256 hash of the method signature.
+
+For example, the method selector for a method named `add` which takes
+two uint64 parameters and returns a uint128 can be computed as
+follows:
+
+```
+Method signature: add(uint64,uint64)uint128
+SHA-512/256 hash (in hex): 8aa3b61f0f1965c3a1cbfa91d46b24e54c67270184ff89dc114e877b1753254a
+Method selector (in hex): 8aa3b61f
+```
+
+
+#### Method Description
+
+A method description provides further information about a method
+beyond its signature. This description is encoded in JSON and consists
+of a method's name, description, arguments (their types, names, and
+descriptions), and return type (and possible description for the
+return type). From this structure, the method's signature and selector
+can be calculated. The Algorand SDKs will provide convenience
+functions to calculate signatures and selectors from such JSON files.
+
+These details will enable high-level languages and dapps/wallets to
+properly encode arguments and call methods, as well as offer users a
+rich description for a method and its arguments. This description can
+populate UIs in dapps, wallets, and explorers with description of
+parameters, as well as populate information about methods in IDEs for
+HLLs.
+
+The JSON structure for such an object is:
+
+```
+interface Method {
+  name: string,
+  desc?: string,
+  args: Array<{ name: string, type: string, desc?: string }>,
+  returns?: { type: string, desc?: string }
+}
+```
+
```

Comment:
If we have a method that also depends on additional application call arguments such as `foreign_assets` or `foreign_apps`, should they be included in the interface definition?

Reply:
Added a section describing how to handle them.  If you use the special types `account`, `asset`, or `application` then the ABI now says to put them into the foreign arrays, and "encode them" with just one byte index into the array.